### PR TITLE
Hiding input when the queue is full #DS-730

### PR DESCRIPTION
**Pull request by tomassychra**
# FileUploader - Hiding input when the queue is full

The branch is based on main, but there are two branches rebased with modifications #817 and #825 

* Web package ✅
* Twig package ✅
* React package ✅

### CODE REVIEW

**Comment by tomassychra** [comment]
I also discovered an error when input is set as multiple and more files than allowed are selected. So I added the same counter as applied for Drag and Drop. Applies to Web and React package.

File: packages/web-react/src/components/FileUploader/useFileUploaderInput.ts [hunk 1]
```diff
@@ -143,7 +167,14 @@ export const useFileUploaderInput = (props: UseFileUploaderInputProps): UseFileU
     onDrop: onDropHandler,
   });
 
+  useEffect(() => {
+    updateDropZoneVisibility();
+    // We only want a dependency for fileQueue
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [fileQueue]);
```
Comment (literat) [inline packages/web-react/src/components/FileUploader/useFileUploaderInput.ts:174]: If you are using this dependency, so pass it to the updater -> `updateDropZoneVisibility(fileQueue)`.

File: packages/web/src/js/FileUploader.ts [hunk 2]
```diff
@@ -198,19 +252,26 @@ class FileUploader extends BaseComponent {
...
 
   onChange(event: Event & { target: HTMLInputElement }) {
     const { target } = event;
     const filesArray = target.files ? [...target.files] : [];
 
-    filesArray.forEach((file) => this.addToQueue(file));
+    let counter = 0;
+    counter += this.fileQueue.size;
+
+    filesArray.forEach((file) => {
```
Comment (literat) [inline packages/web/src/js/FileUploader.ts:266]: Same comment as I have in React about breaking the loop.

**Review by literat — CHANGES_REQUESTED** [review]
Seems fine. Just a few suggestions for the logic.

File: packages/web/src/scss/components/FileUploader/index.html [hunk 3]
```diff
@@ -99,6 +99,113 @@ <h3 id="attachments-multipleFile" hidden>Attachments</h3>
...
+
+    <!-- FileUploaderInput: start -->
+    <div class="FileUploaderInput" data-spirit-element="wrapper">
+      <label for="fileUploadWithDismissible" class="FileUploaderInput__label">Label</label>
+      <input type="file" id="fileUploadWithDismissible" name="attachment1" class="FileUploaderInput__input" multiple data-spirit-element="input" />
+      <div class="FileUploaderInput__dropZone" data-spirit-element="dropZone">
+        <svg width="24" height="24" aria-hidden="true">
+          <use xlink:href="/icons/svg/sprite.svg#upload" />
+        </svg>
+        <label for="fileUpload" class="FileUploaderInput__dropZoneLabel">
```
Comment (literat) [inline packages/web/src/scss/components/FileUploader/index.html:192]: Here there is a bad `for` identification. After clicking on the field I end up with the first file upload on the page, but not this one. Please, fix this.

File: packages/web/src/js/__tests__/FileUploader.test.ts [hunk 4]
```diff
@@ -96,4 +97,80 @@ describe('FileUploader', () => {
...
+      fileUploader = new FileUploader(SelectorEngine.findOne('.file-uploader')!);
+    });
+
+    afterEach(() => {
+      jest.clearAllMocks();
+    });
+
+    it('should hides the drop zone when the file queue limit is reached', async () => {
+      const dropZone = fixtureEl.querySelector('[data-spirit-element="dropZone"]') as HTMLElement;
+      const file1 = new File([''], 'test1.txt', { type: 'text/plain', lastModified: Date.now() });
```
Comment (literat) [inline packages/web/src/js/__tests__/FileUploader.test.ts:160]: Can we please end with using actual date in tests, when it is not necessary? I know that we are already using this somewhere but actually, this is a bad habit to continue with this.

As I mentioned earlier, this can cause some errors in the future when we start working with `lastModified`. Another point is that the test is slightly more performant when they do not need to recalculate actual DateTime.

File: packages/web-react/src/components/FileUploader/useFileUploaderInput.ts [hunk 5]
```diff
@@ -132,8 +150,14 @@ export const useFileUploaderInput = (props: UseFileUploaderInputProps): UseFileU
     const { files } = event.target;
     const filesArray = Array.prototype.slice.call(files);
 
+    let counter = 0;
+    counter += fileQueue.size;
+
     filesArray.forEach((file: File) => {
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/useFileUploaderInput.ts:156]: This is how it works for the drag and drop as well, I just used it for the normal selection when I discovered the bug

File: packages/web/src/js/FileUploader.ts (hunk 2 above)
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:266]: This is how it works for the drag and drop as well, I just used it for the normal selection when I discovered the bug

File: packages/web/src/js/__tests__/FileUploader.test.ts (hunk 4 above)
Comment (tomassychra) [reply packages/web/src/js/__tests__/FileUploader.test.ts:160]: I created issue for that https://jira.lmc.cz/browse/DS-769. I spend too much time for creating unit tests. So that we don't stop ourselves anymore

File: packages/web/src/scss/components/FileUploader/index.html (hunk 3 above)
Comment (tomassychra) [reply packages/web/src/scss/components/FileUploader/index.html:192]: fixed

File: packages/web/src/js/__tests__/FileUploader.test.ts (hunk 4 above)
Comment (tomassychra) [reply packages/web/src/js/__tests__/FileUploader.test.ts:160]: done

File: packages/web/src/js/FileUploader.ts (hunk 2 above)
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:266]: done

File: packages/web-react/src/components/FileUploader/useFileUploaderInput.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/useFileUploaderInput.ts:156]: done

File: packages/web-react/src/components/FileUploader/useFileUploaderInput.ts [hunk 6]
```diff
@@ -76,6 +84,16 @@ export const useFileUploaderInput = (props: UseFileUploaderInputProps): UseFileU
     }
   };
 
+  const updateDropZoneVisibility = () => {
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/useFileUploaderInput.ts:87]: fixed

File: packages/web-react/src/components/FileUploader/useFileUploaderInput.ts (hunk 1 above)
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/useFileUploaderInput.ts:174]: fixed

File: packages/web-react/src/components/FileUploader/README.md [hunk 7]
```diff
@@ -293,6 +310,7 @@ The rest of the properties are created from the default `<div>` element. [More a
 | `maxFileSize`      | `number`                             | 1000000 | ✕        | The maximum size of the uploaded file                                                                                                                           |
 | `maxUploadedFiles` | `number`                             | 10      | ✕        | Maximum file upload queue size                                                                                                                                  |
 | `isMultiple`       | `boolean`                            | -       | ✕        | When multiple files can be selected at once                                                                                                                     |
+| `isDismissible`    | `boolean`                            | -       | ✕        | When to hide FileUploaderInput after reaching file limit in `maxUploadedFiles`                                                                                  |
```
Comment (adamkudrna) [inline packages/web-react/src/components/FileUploader/README.md:313]: How about something more extensible:

```suggestion
| `queueLimitBehavior`    | `disable`, `hide` | `hide` | ✕        | Behavior of FileUploaderInput after reaching file limit in `maxUploadedFiles`                                                                                  |
```

1. We can add more options in the future.
2. IMO, [_dismiss_](https://slovnik.seznam.cz/preklad/anglicky_cesky/dismiss) is not the right word for this… 🤔

File: packages/web-react/src/components/FileUploader/README.md [hunk 8]
```diff
@@ -347,30 +365,31 @@ and a collection of props selected above for easier implementation. Other props
...
+| Prop name             | Type                                    | Default | Required | Description                                                                    |
+| --------------------- | --------------------------------------- | ------- | -------- | ------------------------------------------------------------------------------ |
+| `id`                  | `string`                                | -       | ✔        | UncontrolledFileUploader id                                                    |
+| `inputId`             | `string`                                | -       | ✔        | FileUploaderInput id                                                           |
+| `inputLabel`          | `string`                                | -       | ✔        | FileUploaderInput label                                                        |
+| `inputName`           | `string`                                | -       | ✔        | FileUploaderInput name                                                         |
+| `inputProps`          | `Partial<FileUploaderInputBaseProps>`   | -       | ✕        | Rest of FileUploaderInput props                                                |
```
Comment (adamkudrna) [inline packages/web-react/src/components/FileUploader/README.md:374]: Oh, now I noticed 🙈. When a props is not required, it has a default value (`''`, `null`, `false`, … see the [`Button` `README`](https://github.com/lmc-eu/spirit-design-system/blob/main/packages/web-react/src/components/Button/README.md#available-props), for example) that needs to be listed here in the API table.

Now when we are on it, if you have time, would please fix it? 🙏🏻  But we can do it later, no pressure.

cc @literat

File: packages/web-react/src/components/FileUploader/useFileUploaderInput.ts [hunk 9]
```diff
@@ -109,21 +123,47 @@ export const useFileUploaderInput = (props: UseFileUploaderInputProps): UseFileU
...
 
     if (event.dataTransfer.items) {
-      transferItems.forEach((item) => {
-        if (item.kind === 'file') {
-          const file = item.getAsFile();
-          if (file && counter < maxUploadedFiles) {
-            fileProcessHandler(file);
-            counter += 1;
-          }
+      // transferItems.forEach((item) => {
```
Comment (adamkudrna) [inline packages/web-react/src/components/FileUploader/useFileUploaderInput.ts:126]: The code commented out is to be uncommented in an upcoming PR, I suppose…?

File: packages/web/src/js/FileUploader.ts [hunk 10]
```diff
@@ -23,8 +24,10 @@ class FileUploader extends BaseComponent {
   isDragAndDropSupported: boolean;
   fileSizeLimit: number;
   fileQueueLimit: number;
+  isDismissible: boolean;
   inputName: string;
   isMultiple: boolean;
+  accept: string;
```
Comment (adamkudrna) [inline packages/web/src/js/FileUploader.ts:30]: I think the object properties should be listed alphabetically. (ESlint, can you hear me…? 🙉)

File: packages/web/src/scss/components/FileUploader/README.md [hunk 11]
```diff
@@ -61,6 +61,22 @@ The maximum size of the uploaded file that is validated by the JavaScript plugin
 </div>
 ```
 
+### Dismissible Drop Zone
```
Comment (adamkudrna) [inline packages/web/src/scss/components/FileUploader/README.md:64]: Please update this section accordingly once we resolve the prop name.

File: packages/web/src/scss/components/FileUploader/index.html [hunk 12]
```diff
@@ -99,6 +99,113 @@ <h3 id="attachments-multipleFile" hidden>Attachments</h3>
...
+  </div>
+  <!-- FileUploader: end -->
+</section>
+<section class="docs-Section">
+
+  <h2 class="docs-Heading">Input multiple with Dismissible Feature</h2>
+
+  <p>
+    Input is hidden after the "<code>data-file-queue-limit</code>" limit is reached and "<code>data-spirit-dismissible</code>" is set.
+    After removing the file from the queue, the input will be restored.
```
Comment (adamkudrna) [inline packages/web/src/scss/components/FileUploader/index.html:158]:
```suggestion
    After removing a file from the queue, the input will be restored.
```

This is one of the few places where the article matters quite a lot 🙂.

File: packages/web-react/src/components/FileUploader/stories/FileUploaderDismissible.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    >
+      <FileUploaderInput
+        id="fileUploaderExampleDismissibleInput"
+        name="attachments"
+        label="Label"
+        linkText="Upload your file(s)"
+        labelText="or drag and drop here"
+        helperText="Max file size is 10 MB"
+        validationText="Validation message"
+        maxUploadedFiles={3}
```
Comment (adamkudrna) [inline packages/web-react/src/components/FileUploader/stories/FileUploaderDismissible.tsx:34]: Please mention the file number limit in the helper text so we can see what to test in the demo.

image

Comment (adamkudrna) [inline packages/web-react/src/components/FileUploader/stories/FileUploaderDismissible.tsx:34]: Files should be limited to 3 but I succeeded with selecting 4 files:

image

**Review by adamkudrna — COMMENTED** [review]
Aside from not being sure if the limit works correctly, I only have comments on the name of the new API option.

File: packages/web-react/src/components/FileUploader/README.md [hunk 14]
```diff
@@ -75,6 +75,23 @@ const { fileQueue, addToQueue, clearQueue, onDismiss } = useFileQueue();
 </FileUploader>
 ```
 
+### Dismissible Input
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/README.md:78]: The task is about dismissible, not about file queue

File: packages/web-react/src/components/FileUploader/README.md (hunk 8 above)
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/README.md:374]: But that's wrong. Not always is default value `null` or `false`. Depends on type. For example, the unspecified value for children is not `null` but `undefined`

File: packages/web-react/src/components/FileUploader/stories/FileUploaderDismissible.tsx (hunk 13 above)
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/stories/FileUploaderDismissible.tsx:34]: This is already fixed now

File: packages/web/src/js/FileUploader.ts (hunk 10 above)
Comment (literat) [reply packages/web/src/js/FileUploader.ts:30]: https://www.npmjs.com/package/eslint-plugin-sort-class-members

Comment (literat) [reply packages/web/src/js/FileUploader.ts:30]: I have filed an issue for this.

File: packages/web/src/scss/components/FileUploader/index.html (hunk 12 above)
Comment (tomassychra) [reply packages/web/src/scss/components/FileUploader/index.html:158]: fixed

File: packages/web-react/src/components/FileUploader/README.md (hunk 7 above)
Comment (literat) [reply packages/web-react/src/components/FileUploader/README.md:314]: I have asked ChatGPT to suggest some other names, here are a few that I consider usable: `visibilityOnLimit`, `visibilityOnLimitReached`, `maxUploadedLimitBehavior`, `uploadLimitBehavior`, and `maxUploadedFilesAction`.

Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/README.md:314]: The default state probably shouldn't be 'hidden'. On the contrary, there should be one more value, perhaps 'none', or better `undefined'. Because the user can not specify `queueLimitBehavior` at all and nothing will happen but throw errors. But if I set 'hide' or 'disable' it will either be hidden or disabled

Comment (literat) [reply packages/web-react/src/components/FileUploader/README.md:314]: What do you think about the default value for this should be `disabled`? Or it is correct that the actual state (meaning without any visual change after the limit is reached) is ok and then there should be the default value of `undefined`?

Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/README.md]: I set there as default value 'none' for internal states, then it's like `undefined`

File: packages/web/src/scss/components/FileUploader/README.md (hunk 11 above)
Comment (tomassychra) [reply packages/web/src/scss/components/FileUploader/README.md]: done

**Comment by tomassychra** [comment]
I had to move the attribute data deployment from the main wrapper (.FileUploader) to the wrapper (.FileUploaderInput) element for cross-package compatibility

**Comment by tomassychra** [comment]
So I finally added the events to the task, it was a few lines of code when I was already digging into it. It is also described in the documentation

File: packages/web-react/src/components/FileUploader/useFileUploaderStyleProps.ts [hunk 15]
```diff
@@ -69,7 +72,9 @@ export const useFileUploaderStyleProps = (props?: FileUploaderStyleProps): FileU
         link: classNames(fileUploaderInputLinkClass, ...fileUploaderInputLinkUtilityClasses),
         validationText: fileUploaderInputValidationTextClass,
         dropZone: {
-          root: fileUploaderInputDropZoneClass,
+          root: classNames(fileUploaderInputDropZoneClass, {
+            [fileUploaderInputDropZoneHiddenClass]: props?.isDropZoneHidden && props.queueLimitBehavior === 'hide', // TODO: Modify when the disabled state is ready for the entire component
```
Comment (literat) [inline packages/web-react/src/components/FileUploader/useFileUploaderStyleProps.ts:76]: Please add an issue link to this comment.

File: packages/web/src/js/FileUploader.ts [hunk 16]
```diff
@@ -224,8 +251,12 @@ class FileUploader extends BaseComponent {
       this.checkAllowedFileType(file);
       this.checkQueueLimit();
       this.appendToList(file);
+      this.updateDropZoneVisibility();
+      EventHandler.trigger(this.wrapper, EVENT_CHANGE, { fileQueue: this.fileQueue });
     } catch (error) {
+      // eslint-disable-next-line no-console
       console.warn(error);
```
Comment (literat) [inline packages/web/src/js/FileUploader.ts:258]: Now I am not sure if we should log this into the browser console. We cannot turn this off on production and also we are adding the error to the event, which can be handled on demand by the developer.

So I suggest removing the console.warn.

**Review by literat — CHANGES_REQUESTED** [review]
There is some problem in Storybook story: `TypeError: Cannot read properties of undefined (reading 'size')
 `

See https://deploy-preview-831--spirit-design-system-storybook.netlify.app/?path=/docs/components-fileuploader--file-uploader

File: packages/web/src/js/FileUploader.ts [hunk 17]
```diff
@@ -234,14 +265,32 @@ class FileUploader extends BaseComponent {
...
+      }
+    });
+
+    if (overLimit) {
+      this.checkQueueLimit();
+      EventHandler.trigger(this.wrapper, EVENT_ERROR, new Error(errorMessages.errorMaxUploadedFiles));
+    }
 
     setTimeout(() => {
       target.value = '';
```
Comment (literat) [inline packages/web/src/js/FileUploader.ts:296]: event: uploaded

File: packages/web/src/js/FileUploader.ts [hunk 18]
```diff
@@ -224,8 +251,12 @@ class FileUploader extends BaseComponent {
       this.checkAllowedFileType(file);
       this.checkQueueLimit();
       this.appendToList(file);
+      this.updateDropZoneVisibility();
+      EventHandler.trigger(this.wrapper, EVENT_CHANGE, { fileQueue: this.fileQueue });
```
Comment (literat) [inline packages/web/src/js/FileUploader.ts:255]: event: queuedFile

File: packages/web/src/js/FileUploader.ts [hunk 19]
```diff
@@ -234,14 +265,32 @@ class FileUploader extends BaseComponent {
       const itemElement = SelectorEngine.findOne(`li#${name}`);
```
Comment (literat) [inline packages/web/src/js/FileUploader.ts:265]: event: unqueueFile

File: packages/web/src/js/FileUploader.ts [hunk 20]
```diff
@@ -234,14 +265,32 @@ class FileUploader extends BaseComponent {
...
       this.fileQueue.delete(name);
       itemElement?.remove();
+      this.updateDropZoneVisibility();
+      EventHandler.trigger(this.wrapper, EVENT_CHANGE, { fileQueue: this.fileQueue });
```
Comment (literat) [inline packages/web/src/js/FileUploader.ts:269]: event: unqueuedFile

**Comment by literat** [comment]
I suggest adding these events:
- `upload` -> `uploaded`
- `queueFile` -> `queuedFile`
- `unqueueFile` -> `unqueuedFile`

File: packages/web-react/src/components/FileUploader/useFileUploaderStyleProps.ts (hunk 15 above)
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/useFileUploaderStyleProps.ts:76]: done

File: packages/web/src/js/FileUploader.ts (hunk 16 above)
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:258]: done

File: packages/web/src/scss/components/FileUploader/README.md [hunk 21]
```diff
@@ -329,6 +367,13 @@ const fileList: File[] = myUploaderInstance.getFileQueue();
 const input = myUploaderInstance.inputElement; // Returns an input element, for further use
 ```
 
+### JavaScript Events
+
+| Method                | Description                                                                                                                                   |
+| --------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
+| `change.fileUploader` | This event is fired after every change in the queue. When adding and when removing. Along with the event, the state of the queue is returned. |
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web/src/scss/components/FileUploader/README.md]: done

File: packages/web/src/js/FileUploader.ts [hunk 22]
```diff
@@ -234,14 +265,32 @@ class FileUploader extends BaseComponent {
...
     }
   }
 
   onChange(event: Event & { target: HTMLInputElement }) {
     const { target } = event;
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:277]: done

File: packages/web/src/js/FileUploader.ts [hunk 23]
```diff
@@ -224,8 +251,12 @@ class FileUploader extends BaseComponent {
       this.checkAllowedFileType(file);
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:255]: done

File: packages/web/src/js/FileUploader.ts (hunk 18 above)
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts]: done

File: packages/web/src/js/FileUploader.ts (hunk 20 above)
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts]: done

File: packages/web/src/js/FileUploader.ts (hunk 19 above)
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:268]: done

**Review by literat — APPROVED** [review]
Please change `enqueue` to the `unqueue` and this is ready to go.

File: packages/web/src/scss/components/FileUploader/index.html [hunk 24]
```diff
@@ -149,6 +149,66 @@ <h3 id="attachments-multipleWithAccept" hidden>Attachments</h3>
...
+<section class="docs-Section">
+
+  <h2 class="docs-Heading">FileUploader Multiple with File Queue Behavior Control</h2>
+
+  <p>
+    If FileUploaderInput has a value for `data-spirit-file-queue-limit` set along with `data-spirit-queue-limit-behavior`,
+    the file upload field will be hidden or disabled. When the value for `data-spirit-queue-limit-behavior`
+    is set to 'hide', the input disappears completely. If you set it to 'disable', the input will only be disabled.
+    After removing a file from the queue, the input will be restored.
+    <br />
```
Comment (crishpeen) [inline packages/web/src/scss/components/FileUploader/index.html:161]: Why not a new paragraph? :) We could avoid <br> and it would be semantically better, IMHO.

```suggestion
    </p>
    <p>
```

File: packages/web/src/scss/components/FileUploader/README.md [hunk 25]
```diff
@@ -49,17 +49,8 @@ the available horizontal space:
...
-The maximum size of the uploaded file that is validated by the JavaScript plugin can be adjusted. The default value is
-10 MB. To increase the limit for example to 20 MB, add the `data-max-file-size` attribute:
-
-```html
-<div class="FileUploader" data-toggle="fileUploader" data-max-file-size="20000000">
-  <!-- FileUploaderInput -->
-  <!-- FileUploaderList -->
-</div>
-```
+In this way, the error message will stop being displayed when the limit of the number
```
Comment (crishpeen) [inline packages/web/src/scss/components/FileUploader/README.md:52]: I assume there is an incorrect placement of this sentence. 

Also maybe reword it, so it makes sense. Feel free to consult it in person.

File: packages/web/src/scss/components/FileUploader/README.md [hunk 26]
```diff
@@ -87,6 +78,53 @@ on the root element).
...
+    <!-- ... -->
+  </div>
+  <!-- FileUploaderList -->
+</div>
+```
+
+### Input behavior when the queue is filled (JavaScript)
+
+You can set the input/drop zone to be hidden or disabled when the file queue limit is reached.
+When you set `data-spirit-queue-limit-behavior` together with the desired limit for the queue.
```
Comment (crishpeen) [inline packages/web/src/scss/components/FileUploader/README.md:112]: maybe we could mention here the available options? `hide` and ...?

File: packages/web/src/scss/components/FileUploader/README.md [hunk 27]
```diff
@@ -329,6 +367,16 @@ const fileList: File[] = myUploaderInstance.getFileQueue();
 const input = myUploaderInstance.inputElement; // Returns an input element, for further use
 ```
 
+### JavaScript Events
+
+| Method                      | Description                                                                                                                                 |
+| --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
+| `queueFile.fileUploader`    | This event is fired just before the file was added to the queue.                                                                            |
+| `queuedFile.fileUploader`   | This event is fired after the file was added to queue.                                                                                      |
+| `enqueueFile.fileUploader`  | This event is fired just before the file was removed from the queue.                                                                        |
```
Comment (crishpeen) [inline packages/web/src/scss/components/FileUploader/README.md:376]: doesn't enqueue mean added to the queue? If so, it doesn't make sense? Or maybe I don't understand.

Comment (crishpeen) [reply packages/web/src/scss/components/FileUploader/README.md:376]: Maybe it should be dequeue?

File: packages/web/src/js/__tests__/FileUploader.test.ts [hunk 28]
```diff
@@ -120,17 +122,57 @@ describe('FileUploader', () => {
...
+    `;
+
+      fileUploader = new FileUploader(SelectorEngine.findOne('.file-uploader')!);
+    });
+
+    afterEach(() => {
+      jest.clearAllMocks();
+    });
+
+    it('should hides the drop zone when the file queue limit is reached', async () => {
```
Comment (crishpeen) [inline packages/web/src/js/__tests__/FileUploader.test.ts:160]:
```suggestion
    it('should hide the drop zone when the file queue limit is reached', async () => {
```

File: packages/web/src/scss/components/FileUploader/README.md (hunk 27 above)
Comment (crishpeen) [reply packages/web/src/scss/components/FileUploader/README.md:376]: Ah I see, it is probably a typo? (also in the second enqueue

File: packages/web/src/js/FileUploader.ts [hunk 29]
```diff
@@ -124,12 +144,22 @@ class FileUploader extends BaseComponent {
     }
 
     if (!isTypeSupported) {
-      throw new Error(
-        `The file "${file.name}" is not supported. Please ensure you are uploading a supported file format.`,
-      );
+      throw new Error(`"${file.name}": ${errorMessages.errorFileNotSupported}`);
```
Comment (crishpeen) [inline packages/web/src/js/FileUploader.ts:147]: is it ok we don't throw our own error event too here? I don't know :)

File: packages/web/src/js/FileUploader.ts [hunk 30]
```diff
@@ -14,6 +22,12 @@ const TEMPLATE_ELEMENT_SLOT_NAME = 'data-spirit-populate-field';
 const DATA_DISMISS_ATTRIBUTE = 'data-dismiss';
 const DEFAULT_FILE_SIZE_LIMIT = 10000000; // = 10 MB
 const DEFAULT_FILE_QUEUE_LIMIT = 10;
+const errorMessages = {
+  errorMaxFileSize: 'The file size limit has been exceeded',
+  errorFileDuplicity: 'This file already exists in the queue',
+  errorMaxUploadedFiles: 'You have exceeded the number of files allowed in the queue',
+  errorFileNotSupported: 'is not supported file. Please ensure you are uploading a supported file format.',
```
Comment (crishpeen) [inline packages/web/src/js/FileUploader.ts:29]:
```suggestion
  errorFileNotSupported: 'is not a supported file. Please ensure you are uploading a supported file format.',
```

😁

File: packages/web-twig/src/Resources/components/FileUploader/FileUploaderInput.twig [hunk 31]
```diff
@@ -32,6 +35,9 @@
 {# Attributes #}
 {%- set _nameAttr = _name ? 'name="' ~ _name | escape('html_attr') ~ '"' : null -%}
 {%- set _requiredAttr = _isRequired ? 'required' : null -%}
+{%- set _maxFileSizeAttr = _maxFileSize ? 'data-spirit-max-file-size=' ~ _maxFileSize : null -%}
```
Comment (crishpeen) [inline packages/web-twig/src/Resources/components/FileUploader/FileUploaderInput.twig:38]: I am not sure how we do it in other components, but I have a feeling we prefix these attributes with data if they add data. But please see other components and decide accordingly 
```suggestion
{%- set _dataMaxFileSizeAttr = _maxFileSize ? 'data-spirit-max-file-size=' ~ _maxFileSize : null -%}
```

Comment (crishpeen) [reply packages/web-twig/src/Resources/components/FileUploader/FileUploaderInput.twig:38]: (I am talking only about local variable, the API is ok.)

File: packages/web-react/src/components/FileUploader/useFileUploaderInput.ts [hunk 32]
```diff
@@ -76,6 +80,15 @@ export const useFileUploaderInput = (props: UseFileUploaderInputProps): UseFileU
     }
   };
 
+  const updateDropZoneVisibility = (queue: FileQueueMapType) => {
+    console.log('queueLimitBehavior', queueLimitBehavior);
```
Comment (crishpeen) [inline packages/web-react/src/components/FileUploader/useFileUploaderInput.ts:84]: remove?

**Review by crishpeen — COMMENTED** [review]
I checked mainly the docs and only found small stuff.

File: packages/web/src/js/FileUploader.ts [hunk 33]
```diff
@@ -130,6 +137,16 @@ class FileUploader extends BaseComponent {
     }
   }
 
+  updateDropZoneVisibility() {
+    if (!this.isDismissible) {
+      return;
+    }
+
+    setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:158]: done

File: packages/web/src/scss/components/FileUploader/index.html (hunk 24 above)
Comment (tomassychra) [reply packages/web/src/scss/components/FileUploader/index.html:161]: fixed

File: packages/web/src/scss/components/FileUploader/README.md (hunk 25 above)
Comment (tomassychra) [reply packages/web/src/scss/components/FileUploader/README.md:52]: deleted

File: packages/web/src/scss/components/FileUploader/README.md (hunk 26 above)
Comment (tomassychra) [reply packages/web/src/scss/components/FileUploader/README.md:112]: done

File: packages/web/src/scss/components/FileUploader/README.md (hunk 27 above)
Comment (tomassychra) [reply packages/web/src/scss/components/FileUploader/README.md:376]: fixed

File: packages/web/src/js/__tests__/FileUploader.test.ts (hunk 28 above)
Comment (tomassychra) [reply packages/web/src/js/__tests__/FileUploader.test.ts:160]: fixed

File: packages/web/src/js/FileUploader.ts (hunk 30 above)
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:29]: fixed

File: packages/web/src/js/FileUploader.ts (hunk 29 above)
Comment (tomassychra) [reply packages/web/src/js/FileUploader.ts:147]: it is ok to throw here

File: packages/web-react/src/components/FileUploader/useFileUploaderInput.ts (hunk 32 above)
Comment (tomassychra) [reply packages/web-react/src/components/FileUploader/useFileUploaderInput.ts:84]: fixed
